app/src/main/java/protect/card_locker/importexport/StocardImporter.java

```diff
@@ -121,173 +161,183 @@ public ZIPData importZIP(ZipInputStream zipInputStream, final ZIPData zipData) t
                 }
-            }
-
-            if (startsWith(nameParts, cardBaseName, 1)) {
+            } else if (startsWith(nameParts, cardBaseName, 1)) {
                 // Extract cardName
-                cardName = nameParts[cardBaseName.length].split("\\.", 2)[0];
+                cardName = nameParts[cardBaseName.length];
+
+                StocardRecord record = cards.get(cardName);
+                if (record == null) {
+                    record = new StocardRecord();
+                    cards.put(cardName, record);
+                }
 
                 // This is the card itself
                 if (fileName.endsWith(cardName + "/content.json")) {
                     JSONObject jsonObject = ZipUtils.readJSON(zipInputStream);
-
-                    loyaltyCardHashMap = appendToHashMap(
-                            loyaltyCardHashMap,
-                            cardName,
-                            "cardId",
-                            jsonObject.getString("input_id")
-                    );
+                    record.cardId = jsonObject.getString("input_id");
 
                     if (jsonObject.has("input_provider_name")) {
-                        loyaltyCardHashMap = appendToHashMap(
-                                loyaltyCardHashMap,
-                                cardName,
-                                "store",
-                                jsonObject.getString("input_provider_name")
-                        );
+                        record.store = jsonObject.getString("input_provider_name");
+                    }
+
+                    if (jsonObject.has("label")) {
+                        String label = jsonObject.getString("label");
+                        if (!label.isBlank()) {
+                            record.label = label;
+                        }
                     }
 
                     // Provider ID can be either custom or not, extract whatever version is relevant
-                    String customProviderPrefix = "/users/" + nameParts[1] + "/loyalty-card-custom-providers/";
+                    String customProviderPrefix = "/users/" + userId + "/loyalty-card-custom-providers/";
                     String providerId = jsonObject
                             .getJSONObject("input_provider_reference")
                             .getString("identifier");
                     if (providerId.startsWith(customProviderPrefix)) {
                         providerId = providerId.substring(customProviderPrefix.length());
+                    } else if (providerId.startsWith(PROVIDER_PREFIX)) {
+                        providerId = providerId.substring(PROVIDER_PREFIX.length());
                     } else {
-                        providerId = providerId.substring("/loyalty-card-providers/".length());
+                        throw new FormatException("Unsupported provider ID: " + providerId);
```

Comment:
Hmm, so we're actively changing the importer to fail if the provider can't be found for a card instead of import it with the random ID? If we go that route we probably want to automate updating the provider list to prevent this from getting outdated and users getting unable to import their Stocard data.

Luckily Stocard waits 30 days to give users their export so we can be 30 days late with this lol.

Reply:
No, we're just no longer getting the substring after `len("/loyalty-card-providers/")` without checking if the string actually starts with that prefix. I'm pretty sure that would break if it doesn't. And it never should. Though I guess we could skip those cards instead. Either way there would be no way to find a provider.

Reply:
It will still work correctly for unknown providers. Those are not found and will used the UUID. That's handled in ` importLoyaltyCardHashMap()`. It will just fail if the string doesn't start with either `/loyalty-card-providers/` or `/users/<UUID>/loyalty-card-custom-providers/`. Which seems like the right thing to do.

Reply:
I could change the error message from "Unsupported provider ID" to "Unsupported input_provider_reference identifier format".

Reply:
Because that's what it is. An error when that field has a format we don't support.

Reply:
I didn't really like the idea of silently skipping an unknown format. We want this to be an error so we can add support for it if there's ever a 3rd format.

Reply:
I improved the error message a bit.